node/src/components/consensus/highway_core/highway/vertex.rs

```diff
@@ -53,11 +68,37 @@ where
     C: Context,
 {
     Unit(SignedWireUnit<C>),
+    #[serde(with = "serde_unit_with_panorama")]
+    UnitWithPanorama(SignedWireUnit<C>, Panorama<C>),
     Evidence(Evidence<C>),
     Endorsements(Endorsements<C>),
     Ping(Ping<C>),
 }
 
+/// Serialization and deserialization for a unit with its panorama. This fails to deserialize if
+/// the panorama's hash doesn't match the one in the unit.
+mod serde_unit_with_panorama {
+    use super::*;
+
+    pub(super) fn serialize<S: Serializer, C: Context>(
+        swunit: &SignedWireUnit<C>,
+        panorama: &Panorama<C>,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error> {
+        (swunit, panorama).serialize(serializer)
+    }
+
+    pub(super) fn deserialize<'de, D: Deserializer<'de>, C: Context>(
+        deserializer: D,
+    ) -> Result<(SignedWireUnit<C>, Panorama<C>), D::Error> {
+        let (swunit, panorama) = <(SignedWireUnit<C>, Panorama<C>)>::deserialize(deserializer)?;
+        if panorama.hash() != swunit.wire_unit().panorama_hash {
+            return Err(serde::de::Error::custom("wrong panorama hash"));
```

Comment:
If this is a serialization error, will we be able to ban peers if they give us invalid data? This is domain knowledge and IMO it shouldn't live in (de)serialization logic.

Reply:
I'm not sure: In general, what _do_ we do if we receive data that doesn't deserialize? I think we already report the peer as bad in that case?

Reply:
I definitely think that invariants that can be verified without using any data can be considered deserialization errors, though. This is just a hash that is expected to match the hash of some of the message's data.

Reply:
That rule isn't intrinsic to the bytes/object itself though – the message with an invalid hash can still correctly deserialize to the expected structure. To me, that rule is a domain rule and (de)serialization logic shouldn't know anything about domain, that' all.

Reply:
OK, I'll change it then, and do the check in `pre_validate_vertex` instead.

Reply:
That's only my opinion (only I feel pretty strong about it) and you don't have to do it if you strongly disagree.

Reply:
I only disagree weakly, so I'll do it. :smile: 

I'd _like_ to make it impossible to create an _instance_ of `Vertex` that violates the invariant, and that would require us to make it part of the deserialization logic.
However, we already instantiate by directly calling `UnitWithPanorama(unit, panorama)` in the code anyway — an enum can only enfoce its invariants if it is made private, and wrapped in some other type.

Reply:
If we had a separate representation of each struct for inter-node communication, then the mapping between the two could return errors. The constructor of `Vertex` is the place where _I_ would put these validation steps.

Reply:
> The constructor of `Vertex`

That's what I mean: An enum always has canonical constructors (its variant names), so we can't add validation there unless we wrap the whole enum in another type.

Also, if you do validation of a type in all constructors, you _should_ also do the same validation in the deserialization code, otherwise you create a way to instantiate surprising invalid objects.

Reply:
If there's no way to deserialize directly into an enum, and there's a single constructor that includes the domain validity rules then there's no way to create an instance of an invalid object. In that setup you would need an intermediate DTO that has a some mapping to your domain types and that DTO could be invalid from the POV of the domain but it's OK – b/c DTO is not a domain _yet_.

Reply:
Sure, but as long as the enum isn't private (to its module) and hidden inside another data type, you can always create (outside its module) an invalid instance by directly creating a particular variant.

Reply:
Sure, our job is to write the best code we can and even after making it defensive as much as possible, a malicious developer tries to work around the guards we placed and hurts himself – we can't do anything about it.

Reply:
Done in https://github.com/casper-network/casper-node/pull/1550/commits/bda70086de3433cf17a4a8341138fce929fafd12.